File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/interactors/DecksInteractorImpl.java
package com.dbottillo.mtgsearchfree.interactors;

import android.net.Uri;

import com.dbottillo.mtgsearchfree.model.Deck;
import com.dbottillo.mtgsearchfree.model.MTGCard;
import com.dbottillo.mtgsearchfree.model.storage.DecksStorage;
import com.dbottillo.mtgsearchfree.util.FileUtil;
import com.dbottillo.mtgsearchfree.util.LOG;
import com.dbottillo.mtgsearchfree.util.Logger;

import java.util.List;

import javax.inject.Inject;

import rx.Observable;

public class DecksInteractorImpl implements DecksInteractor {

    private final DecksStorage storage;
    private final FileUtil fileUtil;
    private final Logger logger;

    @Inject
    public DecksInteractorImpl(DecksStorage storage, FileUtil fileUtil, Logger logger) {
        this.logger = logger;
        this.storage = storage;
        this.fileUtil = fileUtil;
        logger.d("created");
    }

    public Observable<List<Deck>> load() {
        logger.d("loadSet decks");
        return Observable.just(storage.load());
    }

    @Override
    public Observable<List<MTGCard>> loadDeck(Deck deck) {
        logger.d("loadSet " + deck.toString());
        return Observable.just(storage.loadDeck(deck));
    }

    @Override
    public Observable<List<Deck>> addDeck(String name) {
        logger.d("create deck with name: " + name);
        return Observable.just(storage.addDeck(name));
    }

    @Override
    public Observable<List<Deck>> deleteDeck(Deck deck) {
        logger.d("delete " + deck.toString());
        return Observable.just(storage.deleteDeck(deck));
    }

    @Override
    public Observable<List<MTGCard>> editDeck(Deck deck, String name) {
        logger.d("edit " + deck.toString() + " with name: " + name);
        return Observable.just(storage.editDeck(deck, name));
    }

    @Override
    public Observable<List<MTGCard>> addCard(Deck deck, MTGCard card, int quantity) {
        logger.d("add " + quantity + " " + card.toString() + " to deck: " + deck);
        return Observable.just(storage.addCard(deck, card, quantity));
    }

    @Override
    public Observable<List<MTGCard>> addCard(String name, MTGCard card, int quantity) {
        logger.d("add " + quantity + " " + card.toString() + " to new deck with name: " + name);
        return Observable.just(storage.addCard(name, card, quantity));
    }

    @Override
    public Observable<List<MTGCard>> removeCard(Deck deck, MTGCard card) {
        logger.d("remove " + card.toString() + " from deck: " + deck);
        return Observable.just(storage.removeCard(deck, card));
    }

    @Override
    public Observable<List<MTGCard>> removeAllCard(Deck deck, MTGCard card) {
        logger.d("remove all " + card.toString() + " from deck: " + deck);
        return Observable.just(storage.removeAllCard(deck, card));
    }

    @Override
    public Observable<List<Deck>> importDeck(Uri uri) {
        logger.d("import " + uri.toString());
        try {
            return Observable.just(storage.importDeck(uri));
        } catch (Throwable throwable) {
            return Observable.error(throwable);
        }
    }

    @Override
    public Observable<Boolean> exportDeck(Deck deck, List<MTGCard> cards) {
        return Observable.just(fileUtil.downloadDeckToSdCard(deck, cards));
    }

    @Override
    public Observable<List<MTGCard>> moveCardToSideboard(Deck deck, MTGCard card, int quantity) {
        logger.d("move " + card.toString() + " to sideboard deck: " + deck);
        return Observable.just(storage.moveCardToSideboard(deck, card, quantity));
    }

    @Override
    public Observable<List<MTGCard>> moveCardFromSideboard(Deck deck, MTGCard card, int quantity) {
        logger.d("move " + card.toString() + " from sideboard deck: " + deck);
        return Observable.just(storage.moveCardFromSideboard(deck, card, quantity));
    }

}

File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/interactors/DecksInteractorImpl.kt
package com.dbottillo.mtgsearchfree.interactors

import android.net.Uri
import com.dbottillo.mtgsearchfree.model.CardsCollection
import com.dbottillo.mtgsearchfree.model.Deck
import com.dbottillo.mtgsearchfree.model.DeckCollection
import com.dbottillo.mtgsearchfree.model.MTGCard
import com.dbottillo.mtgsearchfree.model.storage.DecksStorage
import com.dbottillo.mtgsearchfree.util.FileUtil
import com.dbottillo.mtgsearchfree.util.Logger
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import javax.inject.Inject

class DecksInteractorImpl @Inject
constructor(val storage: DecksStorage,
            val fileUtil: FileUtil,
            val schedulerProvider: SchedulerProvider,
            val logger: Logger) : DecksInteractor {

    init {
        logger.d("created")
    }

    override fun load(): Single<List<Deck>> {
        logger.d("loadSet decks")
        return Single.fromCallable { storage.load() }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun loadDeck(deck: Deck): Observable<DeckCollection> {
        logger.d("loadSet " + deck.toString())
        return Observable.fromCallable { storage.loadDeck(deck) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun copy(deck: Deck): Single<List<Deck>> {
        logger.d("copy deck")
        return Single.fromCallable { storage.copy(deck) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun addDeck(name: String): Observable<List<Deck>> {
        logger.d("create deck with name: " + name)
        return Observable.fromCallable { storage.addDeck(name) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun deleteDeck(deck: Deck): Observable<List<Deck>> {
        logger.d("delete " + deck.toString())
        return Observable.fromCallable { storage.deleteDeck(deck) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun editDeck(deck: Deck, name: String): Single<Deck> {
        logger.d("edit " + deck.toString() + " with name: " + name)
        return Single.fromCallable { storage.editDeck(deck, name) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun addCard(deck: Deck, card: MTGCard, quantity: Int): Observable<DeckCollection> {
        logger.d("add " + quantity + " " + card.toString() + " to deck: " + deck)
        return Observable.fromCallable { storage.addCard(deck, card, quantity) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun addCard(name: String, card: MTGCard, quantity: Int): Observable<DeckCollection> {
        logger.d("add " + quantity + " " + card.toString() + " to new deck with name: " + name)
        return Observable.fromCallable { storage.addCard(name, card, quantity) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun removeCard(deck: Deck, card: MTGCard): Observable<DeckCollection> {
        logger.d("remove " + card.toString() + " from deck: " + deck)
        return Observable.fromCallable { storage.removeCard(deck, card) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun removeAllCard(deck: Deck, card: MTGCard): Observable<DeckCollection> {
        logger.d("remove all " + card.toString() + " from deck: " + deck)
        return Observable.fromCallable { storage.removeAllCard(deck, card) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun importDeck(uri: Uri): Observable<List<Deck>> {
        logger.d("import " + uri.toString())
        try {
            return Observable.fromCallable { storage.importDeck(uri) }
                    .subscribeOn(schedulerProvider.io())
                    .observeOn(schedulerProvider.ui())
        } catch (throwable: Throwable) {
            return Observable.error<List<Deck>>(throwable)
        }
    }

    override fun exportDeck(deck: Deck): Completable {
        return Completable.fromCallable {
            val cards = storage.loadDeck(deck).allCards()
            val exported = fileUtil.downloadDeckToSdCard(deck, CardsCollection(cards, null, true))
            if (exported){
                Completable.complete()
            } else {
                Completable.error(Throwable("deck not exported"))
            }
        }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun moveCardToSideboard(deck: Deck, card: MTGCard, quantity: Int): Observable<DeckCollection> {
        logger.d("move " + card.toString() + " to sideboard deck: " + deck)
        return Observable.fromCallable { storage.moveCardToSideboard(deck, card, quantity) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

    override fun moveCardFromSideboard(deck: Deck, card: MTGCard, quantity: Int): Observable<DeckCollection> {
        logger.d("move " + card.toString() + " from sideboard deck: " + deck)
        return Observable.fromCallable { storage.moveCardFromSideboard(deck, card, quantity) }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
    }

}
